### DISCUSSION TITLE
Progress on Qvvs and Psyshock for 0.7

### DISCUSSION THREAD

**Discussion by laicasaane** [Dreaming381/Latios-Framework#26] [General]
Hi, I'm eager to see the concrete shape of Qvvs and Psyshock modules. I'm using Transform V2 and wanting to experiment with Psyshock but it is currently depending on Transform V1. So I want to ask if you have an ETA for 0.7 release so I can plan my project roadmap better?

**Comment by Dreaming381** [comment]
Hi! Currently Transforms and Psyshock are fully written but completely untested. I have the renderer torn apart right now as I am overhauling the skinning system to support a boatload of new features. That will probably take me all month yet. But after that, I will be entering a debugging stage where I test everything out and iron out all the bugs. Once I have LSSS working again, I will enter a public beta stage while I iron out Kinemation bugs and put together all the documentation. Actual release will likely be April at this point.

If you'd like, I can put together a zip of Core, Transforms, and Psyshock. That way, you'd at least be able to try out Psyshock's new static APIs. As long as you don't actually install Transforms into the bootstrap and stick to only using NativeArray<ColliderBody> when building CollisionLayers, you would have nearly everything Psyshock has to offer in a V2 environment. While I won't rule out regressions from some stupid typo or two, most of the rest of Psyshock changes were relatively mild. Mostly just reorganizing the code, making the old API internal, and making a new Qvvs frontend.

Is that something you would like?

**Reply by laicasaane** [reply]
Thanks. I would like to have the zip of Core, Transforms and Psyshock. My project is a 2D game using NSprites to render spritesheet animations so I don't need Kinemation after all.

**Reply by Dreaming381** [reply]
[latios-framework-07-preview0.zip](https://github.com/Dreaming381/Latios-Framework/files/10613906/latios-framework-07-preview0.zip)
If you have any questions or encounter some unexpected behavior, do not hesitate to reach out to me!

**Reply by laicasaane** [reply]
Thanks! I'll surely report if any.

**Reply by laicasaane** [reply]
@Dreaming381 There is a problem with SmartBlobber when I use ICustomBootstrap

```
[Worker0] InvalidOperationException: Group of type Latios.Authoring.Systems.SmartBlobberCleanupBakingGroup has not been created, either the derived class forgot to call base.OnCreate(), or it has been destroyed
```

The problem starts here at `group.AddSystemToUpdateList(system);`
```cs
public partial struct SmartBlobberTools<TBlobType> where TBlobType : unmanaged
{
    public void Register(World managedWorld)
    {
        if (managedWorld.GetExistingSystemManaged<Systems.SmartBlobberTypedPostProcessBakingSystem<TBlobType> >() != null)
                return;
            var system = managedWorld.GetOrCreateSystemManaged<Systems.SmartBlobberTypedPostProcessBakingSystem<TBlobType> >();
            var group  = managedWorld.GetExistingSystemManaged<Systems.SmartBlobberCleanupBakingGroup>();
            group.AddSystemToUpdateList(system);
    }
}
```

Note: This problem only happens after I imported the content of your zip.

**Reply by Dreaming381** [reply]
I just tried reproducing this in an empty project. While there were some compiler errors with some lingering Transforms V1 stuff in Core that didn't need to be there (I'm assuming you just deleted that code as well), I was otherwise unable to reproduce the issue. Subscene baking is working for me.

Could you show me what your bootstrap looks like and give me more details with your error messages? Can you get the error to occur while opening a subscene? That will give you a more detailed stack trace.

**Reply by laicasaane** [reply]
I've fixed it. It was my mistake when I changed the namespace of Psyshock but didn't do that thoroughly.

**Comment by laicasaane** [comment]
@Dreaming381 Is it possible to create InstantiateCommandBuffers for up to 15 components? Based on the current implementation I think we can use T4 to generate the rest of them. Could you give me some insights about this because I have a need to initialize more than 5 components at instantiation.

**Reply by Dreaming381** [reply]
Yeah. It should be possible to make that happen now that `ComponentTypeSet` can be initialized with `FixedList128Bytes<ComponentType>`.
If you made an attempt at it, I would definitely be willing to integrate such changes.
Aside from InstantiateCommandBuffer.cs which you would probably want to generate with T4, you will also need to modify this file: https://github.com/Dreaming381/Latios-Framework/blob/v0.6.4/Core/Internal/InstantiateCommandBufferUntyped.cs
There are Add methods for both normal and ParallelWriter in there. It also has a WriteComponentDataJob which needs to account for the number of component types and use dummy types for unused type handles.

Let me know if you need a deeper explanation as to what is happening in that code, or if you have any other questions!

**Reply by laicasaane** [reply]
Today I decide to proceed with this codegen. So I want to ask why we need `FixedList128Bytes<ComponentType>`?

> Let me know if you need a deeper explanation as to what is happening in that code, or if you have any other questions!

Yes, it would be much helpful if you point out to me where should I be aware of.

**Reply by Dreaming381** [reply]
`InstantiateCommandBuffer` is designed to add the components it initializes to the entities if they don't already exist on the prefab. It does so in bulk operations using `ComponentTypeSet` for performance reasons. `ComponentTypeSet` only has constructors for up to 5 `ComponentType` instances. Anything more than that, and you either have to give it a managed array (not allowed in a Bursted `ISystem`), or give it a `FixedList128Bytes<ComponentType>`.

In an earlier release, `ComponentTypeSet` didn't have that `FixedList128Bytes<ComponentType>` constructor, which is why I stopped at supporting 5 components. `ComponentType` is 8 bytes in size, so `FixedList64Bytes<ComponentType>` can only hold 7 component types, whereas `FixedList128Bytes<ComponentType>` can hold 15. `InstantiateCommandBufferUntyped` takes in a `FixedList64Bytes<ComponentType>`, so you will need to modify that to `FixedList128Bytes<ComponentType>` and change the `BuildComponentTypesFromFixedList()` method (also legacy code from before that constructor) to work with more types. Or you can delete that method and change the code at the callsites.
